pinot-clients/pinot-java-client/src/main/java/org/apache/pinot/client/BrokerCache.java

```diff
@@ -0,0 +1,165 @@
+    String responseBody = response.getResponseBody(StandardCharsets.UTF_8);
+    return OBJECT_MAPPER.readValue(responseBody, RESPONSE_TYPE_REF);
+  }
+
+  private BrokerData getBrokerData(Map<String, List<BrokerInstance>> responses) {
+    Set<String> brokers = new HashSet<>();
+    Map<String, List<String>> tableToBrokersMap = new HashMap<>();
+    Set<String> uniqueTableNames = new HashSet<>();
+
+    for (Map.Entry<String, List<BrokerInstance>> tableToBrokers : responses.entrySet()) {
+      List<String> brokersForTable = new ArrayList<>();
+      tableToBrokers.getValue().forEach(br -> {
+        String brokerHostPort = br.getHost() + ":" + br.getPort();
+        brokersForTable.add(brokerHostPort);
+        brokers.add(brokerHostPort);
+      });
+      String tableName = tableToBrokers.getKey();
+      tableToBrokersMap.put(tableName, brokersForTable);
+
+      String rawTableName = TableNameBuilder.extractRawTableName(tableName);
+      uniqueTableNames.add(rawTableName);
+    }
+
+    // Add table names without suffixes
+    uniqueTableNames.forEach(tableName -> {
+      if (!tableToBrokersMap.containsKey(tableName)) {
+        // 2 possible scenarios:
+        // 1) Both OFFLINE_SUFFIX and REALTIME_SUFFIX tables present -> use intersection of both the lists
+        // 2) Either OFFLINE_SUFFIX or REALTIME_SUFFIX (and npt both) raw table present -> use the list as it is
+
+        String offlineTable = tableName + ExternalViewReader.OFFLINE_SUFFIX;
+        String realtimeTable = tableName + ExternalViewReader.REALTIME_SUFFIX;
+        if (tableToBrokersMap.containsKey(offlineTable) && tableToBrokersMap.containsKey(realtimeTable)) {
+          List<String> realtimeBrokers = tableToBrokersMap.get(realtimeTable);
+          List<String> offlineBrokers = tableToBrokersMap.get(offlineTable);
+          List<String> tableBrokers = realtimeBrokers
+              .stream()
+              .filter(offlineBrokers::contains)
+              .collect(Collectors.toList());
+          tableToBrokersMap.put(tableName, tableBrokers);
+        } else {
+          tableToBrokersMap.put(tableName,
+              tableToBrokersMap.getOrDefault(offlineTable,
+                  tableToBrokersMap.getOrDefault(realtimeTable, new ArrayList<>())));
+        }
+      }
+    });
+
+    return new BrokerData(tableToBrokersMap, new ArrayList<>(brokers));
+  }
+
+  protected void updateBrokerData()
+      throws Exception {
+    Map<String, List<BrokerInstance>> responses = getTableToBrokersData();
+    _brokerData = getBrokerData(responses);
+  }
+
+  public String getBroker(String tableName) {
+    List<String> brokers = (tableName == null) ? _brokerData._brokers : _brokerData._tableToBrokerMap.get(tableName);
+    return brokers.get(_random.nextInt(brokers.size()));
+  }
+
+  public List<String> getBrokers() {
+    return _brokerData._brokers;
+  }
+
+  public static class BrokerData {
+    public Map<String, List<String>> _tableToBrokerMap;
```

Comment:
Any particular reason for keeping them public? We should be keeping them private and use accessors here.

Reply:
Moved to a new class outside with getters and marked fields as final. My idea was to keep this class as an inner class so only BrokerCache can access it. But I think keeping it separate might be better if in future more classes need to access it.